### PR TITLE
Support `watchedFiles` in `Bun.plugin`

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

WIP. Lets onLoad plugins return a file path to add to watch. This makes it easier to do hot reloading with custom loaders & files.

Example code:

```js
import { plugin } from "bun";
import { compile } from "svelte/compiler";
import { readFileSync } from "fs";

await plugin({
  name: "svelte loader",
  async setup(builder) {
    builder.onLoad({ filter: /\.svelte$/ }, ({ path }) => ({
      contents: compile(readFileSync(path, "utf8"), {
        filename: path,
        generate: "ssr",
      }).js.code,
      loader: "js",
      watchedFilePaths: [path],
      // ^ watch this path
    }));
  },
});
```

If someone else wants to finish this PR, the next steps are roughly:
- Open the file path provided so that we get a file descriptor (for watching on macOS, mostly)
- Call `vm.bun_watcher.addFile`. Set `clone_path` to `true`
- A test which checks that hot-reloading works with e.g. `.svelte` files

I think it'd be fine to only support a single path to ship the initial version. 

### How did you verify your code works?

It doesn't work yet

### CODE REVIEW

File: src/bun.js/bindings/ModuleLoader.cpp [hunk 1]
```diff
@@ -242,6 +254,46 @@ OnLoadResult handleOnLoadResultNotPromise(Zig::GlobalObject* globalObject, JSC::
     result.value.sourceText.value = JSValue {};
     result.value.sourceText.string = {};
 
+    if (Bun__watchMode != HotReload::none) {
+        if (JSC::JSValue watchedFilePathsValue = object->getIfPropertyExists(globalObject, JSC::Identifier::fromString(vm, "watchedFilePaths"_s))) {
```
Comment (Electroid) [inline src/bun.js/bindings/ModuleLoader.cpp:258]: Just `watchedPaths`?

**Review by paperclover — CHANGES_REQUESTED** [review]
why is it called `watchedFilePaths`? this should match esbuild and be called `watchFiles`, and if it is easy then `watchDirs` should also be adopted.

see https://esbuild.github.io/plugins/#on-resolve-arguments

**Comment by paperclover** [comment]
This partially implements #4689 which had two duplicates I closed now.

**Comment by Jarred-Sumner** [comment]
This still needs to be addressed, but it will happen in a different PR

**Pull request closed without merging**